ui/plugins/eu.esdihumboldt.hale.ui/src/eu/esdihumboldt/hale/ui/io/schema/SourceSchemaActionUIAdvisor.java

```diff
@@ -0,0 +1,72 @@
+
+package eu.esdihumboldt.hale.ui.io.schema;
+
+import org.eclipse.ui.PlatformUI;
+
+import eu.esdihumboldt.hale.common.instance.model.DataSet;
+import eu.esdihumboldt.hale.common.instance.model.Instance;
+import eu.esdihumboldt.hale.common.instance.model.InstanceCollection;
+import eu.esdihumboldt.hale.common.instance.model.ResourceIterator;
+import eu.esdihumboldt.hale.common.schema.SchemaSpaceID;
+import eu.esdihumboldt.hale.common.schema.model.Schema;
+import eu.esdihumboldt.hale.ui.io.action.AbstractActionUIAdvisor;
+import eu.esdihumboldt.hale.ui.service.instance.InstanceService;
+import eu.esdihumboldt.hale.ui.service.schema.SchemaService;
+
+/**
+ * Source Schema action UI advisor.
+ * 
+ * @author Kapil Agnihotri.
+ */
+public class SourceSchemaActionUIAdvisor extends AbstractActionUIAdvisor<Schema> {
+
+	/**
+	 * @see eu.esdihumboldt.hale.ui.io.action.ActionUIAdvisor#supportsRemoval(java.lang.String)
+	 */
+	@Override
+	public boolean supportsRemoval(String resourceId) {
+		InstanceService is = PlatformUI.getWorkbench().getService(InstanceService.class);
+		InstanceCollection instances = is.getInstances(DataSet.SOURCE);
+		SchemaService ss = PlatformUI.getWorkbench().getService(SchemaService.class);
+		Schema schema = ss.getSchemas(resourceId, SchemaSpaceID.SOURCE);
+
+		// check if the source schema is associated with any instance, then it
+		// can't be removed.
+		ResourceIterator<Instance> it = instances.iterator();
```

Comment:
Iterating over all instances can be an issue here, which could cause a large delay to actually showing the context menu.

I think it would make more sense to use this method in the instance service: https://github.com/halestudio/hale/blob/76f6c9cc7106cbcb51257028573dd6970aa3dd81/ui/plugins/eu.esdihumboldt.hale.ui/src/eu/esdihumboldt/hale/ui/service/instance/InstanceService.java#L57

Reply:
Code modified to use `getInstanceTypes(DataSet dataset)`.